topdown/eval.go

```diff
@@ -502,7 +502,7 @@ func (e *eval) evalNotPartial(iter evalIterator) error {
 	//	(!A && !C) || (!A && !D) || (!B && !C) || (!B && !D)
 	return complementedCartesianProduct(savedQueries, 0, nil, func(q ast.Body) error {
 		return e.saveInlinedNegatedExprs(q, func() error {
-			return e.next(iter)
+			return iter(e)
```

Comment:
@tsandall any pointers on how to test this one? Looking at the other "product" partial eval tests I wasn't sure how to get one to reproduce the problem. The other two changes are caught by the partial eval check in the normal topdown test cases added. Just this one that I'm not sure about..

Reply:
Something like this will do it:

```
not input with input as 1 with data.x as 1
data.x = 1
```

* `not input` will evaluate and get saved--importantly, the partial evaluation result is small enough to inline.
* `data.x = 1` should fail unless `data.x` is set by the caller--`with data.x as 1` should be reset to the original value by the time the second expression is evaluated